methods overriden: equals and hashCode

yurwing opened the pull request.

yurwing committed:
methods overriden: equals and hashCode

dmytro-hryhoruk commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -22,4 +24,29 @@ public Integer getLength() {
+        }
+        if (rectangle.getClass().equals(Rectangle.class)) {
+            Rectangle rectangle1 = (Rectangle) rectangle;
+            return Objects.equals(getColor(), rectangle1.color)
+                    && Objects.equals(getLength(), rectangle1.length)
+                    && Objects.equals(getWidth(), rectangle1.width);
```
https://mate-academy.github.io/jv-program-common-mistakes/java-core/equals-and-hashcode/equals-and-hashcode

dmytro-hryhoruk commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -22,4 +24,29 @@ public Integer getLength() {
+
+    public int hashCode() {
+        int result = 17;
+        result = 31 * result + (getColor() == null ? 0 : color.hashCode());
```
do don't have to use getter inside your class

yurwing committed:
use primitives directly instead of using getters and replace Objects.equals
methods overriden: equals and hashCode

dmytro-hryhoruk commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -22,4 +22,32 @@ public Integer getLength() {
+        }
+        if (rectangle.getClass().equals(Rectangle.class)) {
+            Rectangle rectangle1 = (Rectangle) rectangle;
+            return (color == rectangle1.color || color.equals(rectangle1.color)
+                    || color == null)
```
NPE when color == null and rectangle1.color is not

dmytro-hryhoruk commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -22,4 +22,32 @@ public Integer getLength() {
+            return false;
+        }
+        if (rectangle.getClass().equals(Rectangle.class)) {
+            Rectangle rectangle1 = (Rectangle) rectangle;
```
don't use numbers in naming

yurwing committed:
get rid of potential NPE and rename "rectangle1" to "comparedRectangle"
use primitives directly instead of using getters and replace Objects.equals
methods overriden: equals and hashCode

dmytro-hryhoruk commented on src/main/java/core/basesyntax/Rectangle.java:
```diff
@@ -22,4 +22,32 @@ public Integer getLength() {
+        }
+        if (rectangle.getClass().equals(Rectangle.class)) {
+            Rectangle comparedRectangle = (Rectangle) rectangle;
+            return (color == comparedRectangle.color || color.equals(comparedRectangle.color)
+                    || (color == null && comparedRectangle.color == null))
```
still NPE:
let's go step by step with such conditions(color = null, comparedRectangle.color = "Red"):
1) First check (color == comparedRectangle.color) => false;
2) Second check (color.equals(comparedRectangle.color)) => NullPointerException because color = null;
3) Third check can't be reached in this case;

yurwing committed:
get rid of potential NPE | attempt 2.0
get rid of potential NPE and rename "rectangle1" to "comparedRectangle"
use primitives directly instead of using getters and replace Objects.equals
methods overriden: equals and hashCode
